### PR TITLE
feat: add git lfs admin option

### PR DESCRIPTION
**Pull request by viceice**
## Changes:

This add a git lfs admin option, which is disabled by default. It suppress pulling lfs objects on clone / fetch / pull.

## Context:

this is a preparation for #6842

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [x] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by viceice** [comment]
@rarkins I think this can be a breaking for some users, if they use git-lfs. Should we enable it by default to not break? or auto enable with a warning when `gitNoVerify` is set but `allowGitLfs` is disabled?

File: docs/usage/self-hosted-configuration.md [hunk 1]
```diff
@@ -11,6 +11,10 @@ Please also see [Self-Hosted Experimental Options](./self-hosted-experimental.md
 
 ## allowCustomCrateRegistries
 
+## allowGitLfs
+
+**Note** You need to also configure `gitNoVerify` to allow commit and push git lfs objects.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:16]: I feel like this text is missing something?

How about:

```suggestion
**Note** You need to also configure `gitNoVerify` to `true` to allow Renovate to commit and push Git LFS objects.
```

I think the correct way to capitalize is `Git LFS` by the way. 😉

File: docs/usage/self-hosted-configuration.md [hunk 2]
```diff
@@ -224,6 +228,8 @@ The flag can be passed to `git commit` and/or `git push`.
 Read the documentation for [git commit --no-verify](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---no-verify) and [git push --no-verify](https://git-scm.com/docs/git-push#Documentation/git-push.txt---no-verify) to learn exactly what each flag does.
 To learn more about Git hooks, read the [Pro Git 2 book, section on Git Hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks).
 
+**Note** If you use git-lfs checkout `allowGitLfs` option.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:231]:
```suggestion
**Note** If you use Git LFS you also need to configure the `allowGitLfs` option to `true`.
```

Maybe something like this is clearer?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
The docs need to explain better what the reader should do to get things working. 😉

File: docs/usage/self-hosted-configuration.md (hunk 1 above)
Comment (viceice) [reply docs/usage/self-hosted-configuration.md:16]: Sure, but `gitNoVerify` is a string array option, so no `true` here

Comment (HonkingGoose) [reply docs/usage/self-hosted-configuration.md:16]: Updated my suggestion. 😄

File: docs/usage/self-hosted-configuration.md [hunk 3]
```diff
@@ -11,6 +11,11 @@ Please also see [Self-Hosted Experimental Options](./self-hosted-experimental.md
 
 ## allowCustomCrateRegistries
 
+## allowGitLfs
+
+**Note** You need to also set the Git commands that Renovate passes to the `gitNoVerify` option.
+This way Renovate can commit and push Git LFS objects.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:17]:
```suggestion
Use this option to allow Renovate to work with repositories which use Git LFS (Git Large File Support).

**Note** You need to also set the Git commands that Renovate passes to the `gitNoVerify` option.
This way Renovate can commit and push Git LFS objects.
```

How about a short introductory sentence, that explains what this option does?

Or do we think the section name is good enough on its own?

File: docs/usage/self-hosted-configuration.md [hunk 4]
```diff
@@ -224,6 +229,8 @@ The flag can be passed to `git commit` and/or `git push`.
 Read the documentation for [git commit --no-verify](https://git-scm.com/docs/git-commit#Documentation/git-commit.txt---no-verify) and [git push --no-verify](https://git-scm.com/docs/git-push#Documentation/git-push.txt---no-verify) to learn exactly what each flag does.
 To learn more about Git hooks, read the [Pro Git 2 book, section on Git Hooks](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks).
 
+**Note** If you use Git LFS you also need to configure the `allowGitLfs` option to `true`.
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:232]:
```suggestion
**Note** If you use Git LFS you also need to set the `allowGitLfs` option to `true`.
```

Maybe `set` is a better word than `configure` here?

File: docs/usage/self-hosted-configuration.md (hunk 3 above)
Comment (viceice) [reply docs/usage/self-hosted-configuration.md:17]: Remember, that the docs compiler will also prepend the message and default values from config definitions file.

**Review by HonkingGoose — DISMISSED** [review]

File: docs/usage/self-hosted-configuration.md [hunk 5]
```diff
@@ -11,6 +11,11 @@ Please also see [Self-Hosted Experimental Options](./self-hosted-experimental.md
 
 ## allowCustomCrateRegistries
 
+## allowGitLfs
+
+**Note** You need to also set the Git commands that Renovate passes to the `gitNoVerify` option.
```
Comment (fgblomqvist) [inline docs/usage/self-hosted-configuration.md:16]: This isn't fully correct. Renovate isn't passing any commands to the `giNoVerify` option. Instead, I would perhaps phrase it something like
> You also need to set the `gitNoVerify` option to the appropriate values for LFS to work.
...ish.

File: lib/config/definitions.ts [hunk 6]
```diff
@@ -7,6 +7,13 @@ import * as pep440Versioning from '../versioning/pep440';
 import type { RenovateOptions } from './types';
```
Comment (fgblomqvist) [inline lib/config/definitions.ts:7]: Could make sense to add a `default` key to the `gitNoVerify` option in this MR, realized that was missing. The default is `['commit', 'push']`.

File: lib/config/definitions.ts [hunk 7]
```diff
@@ -7,6 +7,13 @@ import * as pep440Versioning from '../versioning/pep440';
...
 
 const options: RenovateOptions[] = [
+  {
+    name: 'allowGitLfs',
+    description: 'If true allow git-lfs to pull files.',
+    type: 'boolean',
+    default: false,
```
Comment (fgblomqvist) [inline lib/config/definitions.ts:14]: Per your comment, I think it's better to default this to `true` for now, unless you want to hold it until the next breaking release.

**Review by fgblomqvist — COMMENTED** [review]
I think it'll be fine without any warnings, since the docs do mention that you need to configure `gitNoVerify` as well.

**Comment by viceice** [comment]
I've now implemented all required configuration for git-lfs to work. It's also enabled by default so for most people it's a non breaking.

It's only breaking for self-hosted users who need to override the `gitLfsInclude` option, because it's missing some files.

File: docs/usage/configuration-options.md [hunk 8]
```diff
@@ -705,6 +705,18 @@ If you enabled `automerge` in the Renovate config, you can speed up the automerg
 Caution (fixed in GitLab >= 12.7): when this option is enabled it is possible due to a bug in GitLab that MRs with failing pipelines might still get merged.
 This is caused by a race condition in GitLab's Merge Request API - [read the corresponding issue](https://gitlab.com/gitlab-org/gitlab/issues/26293) for details.
 
+## gitLfsExclude
+
+See also `allowGitLfs` and `gitNoVerify` option.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:710]:
```suggestion
See also the `allowGitLfs` and `gitNoVerify` options.
```

File: docs/usage/configuration-options.md [hunk 9]
```diff
@@ -705,6 +705,18 @@ If you enabled `automerge` in the Renovate config, you can speed up the automerg
...
+
+**Note** This is not supported in WhiteSource Renovate Hosted App.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:712]:
```suggestion
**Note** This is not supported in the WhiteSource Renovate Hosted App.
```

File: docs/usage/configuration-options.md [hunk 10]
```diff
@@ -705,6 +705,18 @@ If you enabled `automerge` in the Renovate config, you can speed up the automerg
...
+
+## gitLfsInclude
+
+See also `allowGitLfs` and `gitNoVerify` option.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:716]:
```suggestion
See also the `allowGitLfs` and `gitNoVerify` options.
```

File: docs/usage/configuration-options.md [hunk 11]
```diff
@@ -705,6 +705,18 @@ If you enabled `automerge` in the Renovate config, you can speed up the automerg
...
+
+**Note** This is not supported in WhiteSource Renovate Hosted App.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:718]:
```suggestion
**Note** This is not supported in the WhiteSource Renovate Hosted App.
```

File: docs/usage/self-hosted-configuration.md [hunk 12]
```diff
@@ -11,6 +11,13 @@ Please also see [Self-Hosted Experimental Options](./self-hosted-experimental.md
...
+This way Renovate can commit and push Git LFS objects.
+
+**Note** This currently don't work in official renovate docker images, see [containerbase/buildpack#13](https://github.com/containerbase/buildpack/issues/13)
```
Comment (HonkingGoose) [inline docs/usage/self-hosted-configuration.md:19]:
```suggestion
**Note** `allowGitLfs` does not work in the official Renovate docker images, see issue [containerbase/buildpack#13](https://github.com/containerbase/buildpack/issues/13) for more information.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]
Some improvements to the wording. 😉

File: docs/usage/configuration-options.md [hunk 13]
```diff
@@ -705,6 +705,13 @@ If you enabled `automerge` in the Renovate config, you can speed up the automerg
 Caution (fixed in GitLab >= 12.7): when this option is enabled it is possible due to a bug in GitLab that MRs with failing pipelines might still get merged.
 This is caused by a race condition in GitLab's Merge Request API - [read the corresponding issue](https://gitlab.com/gitlab-org/gitlab/issues/26293) for details.
 
+## gitLabIgnoreApprovals
+
+Ignore the default project level approval(s), so that Renovate bot can automerge its merge requests, without needing approval(s).
+Under the hood, it creates a MR-level approval rule where `approvals_required` is set to `0`.
+This option works only when `automerge=true`, `automergeType=pr` and `gitLabAutomerge=true`.
```
Comment (HonkingGoose) [inline docs/usage/configuration-options.md:712]:
```suggestion
To use `gitLabIgnoreApprovals` you need to set `automerge=true`, `automergeType=pr` and `gitLabAutomerge=true`.
```

Maybe this is clearer?

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Comment by rarkins** [comment]
@viceice should we close this for now?

**Comment by viceice** [comment]
No, i like to move this forward.

**Comment by steschi** [comment]
Is there a chance for renovate to support git-lfs eventually?

**Comment by rarkins** [comment]
Closing due to staleness

**Pull request closed without merging**